### PR TITLE
`--output-unselected` option in `samtools view`

### PR DESCRIPTION
**Pull request by priyanka-surana**
This PR adds the ability to retrieve the reads/alignments **not** selected by any filters, by adding the `--output-unselected` option (also called `--unoutput`).

## PR checklist

Closes #XXX 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by priyanka-surana** [comment]
Even though CI says all checks have passed, they are not running correctly. I added 4 view statements to the `filter` test but the outputs of only 2 are correct:
```
SAMTOOLS_VIEW.out.bam.view()
// Emits nothing
// Should be [[id:test], /path/to/test.bam]
SAMTOOLS_VIEW.out.csi.view()
// Emits [[id:test], [/path/to/test.bam.csi, /path/to/test.unoutput.bam.csi]]
// Should be [[id:test], /path/to/test.bam.csi]
SAMTOOLS_VIEW.out.unoutput.view()
// Emits [[id:test], /path/to/test.unoutput.bam] as expected 👍 
SAMTOOLS_VIEW.out.unoutput_index.view()
// Emits [[id:test], /path/to/test.unoutput.bam.csi] as expected 👍 
```
Similar issues with other tests as well. All the files are there in the work directory but not being picked up by `emit` statements correctly. Any suggestions?

File: modules/nf-core/samtools/view/meta.yml [hunk 1]
```diff
@@ -68,6 +68,11 @@ output:
       type: file
       description: optional CRAM file index
       pattern: "*.{crai}"
+  # filtered is created when using `--qname-file`
+  - filtered:
+      type: file
+      description: optional file with unselected alignments
+      pattern: "*.filtered.*"
```
Comment (muffato) [inline modules/nf-core/samtools/view/meta.yml:75]: "filtered", "unselectec", or "unoutput" ? Could we settle on just one term ?

**Review by muffato — COMMENTED** [review]
As much as I'd love to know how to make the patterns / regexes work, I'm thinking of a workaround: what about putting the "unoutput" files in a sub-directory ? 
```
def readnames = qname ? "--qname-file ${qname} --unoutput unoutput/${prefix}.${file_type}": ""
```
(there needs to be a `mkdir unoutput` somewhere in the script section)

This way, the original filters like `*.bam` are all still valid, and the unoutput channel would be something like
```
tuple val(meta), path("unoutput/*.{bam,cram,sam}")
```

**Comment by muffato** [comment]
> As much as I'd love to know how to make the patterns / regexes work, I'm thinking of a workaround: what about putting the "unoutput" files in a sub-directory ?

I gave it a try and I didn't like the fact that Nextflow then adds `unoutput` to the publish-dir.

I tried a few things with the syntax `[!unoutput]`. As far as I can tell, the syntax only allows complete file names to be put inside the square brackets, without variables. So my solution (which I've pushed to this branch) is to declare all the outputs in their right channels using complete file names and `${prefix}`.
At the same time, I've renamed `unoutput` / `unfiltered` to `unselected` to match the description of the option ("Output reads not selected by filters to FILE").

**Comment by JoseEspinosa** [comment]
> > As much as I'd love to know how to make the patterns / regexes work, I'm thinking of a workaround: what about putting the "unoutput" files in a sub-directory ?
> 
> I gave it a try and I didn't like the fact that Nextflow then adds `unoutput` to the publish-dir.
> 
> I tried a few things with the syntax `[!unoutput]`. As far as I can tell, the syntax only allows complete file names to be put inside the square brackets, without variables. So my solution (which I've pushed to this branch) is to declare all the outputs in their right channels using complete file names and `${prefix}`. At the same time, I've renamed `unoutput` / `unfiltered` to `unselected` to match the description of the option ("Output reads not selected by filters to FILE").

Maybe late to the party but won't the `**` pattern be useful to just recover files (without adding `unoutput` to the publish-dir). See [here](https://www.nextflow.io/docs/latest/process.html#multiple-output-files): "Directories are included, unless the ** pattern is used to recurse through directories".

**Comment by muffato** [comment]
@JoseEspinosa : you mean something like this ?
```
    tuple val(meta), path("**.bam"),                                   emit: bam,            optional: true
    tuple val(meta), path("**.cram"),                                  emit: cram,           optional: true
    tuple val(meta), path("**.sam"),                                   emit: sam,            optional: true
    tuple val(meta), path("**.${file_type}.bai"),                      emit: bai,            optional: true
    tuple val(meta), path("**.${file_type}.csi"),                      emit: csi,            optional: true
    tuple val(meta), path("**.${file_type}.crai"),                     emit: crai,           optional: true
    tuple val(meta), path("*.unselected.${file_type}"),                emit: unselected,       optional: true
    tuple val(meta), path("*.unselected.${file_type}.{bai,csi,crsi}"), emit: unselected_index, optional: true
    path  "versions.yml",                                              emit: versions
```

**Review by maxulysse — DISMISSED** [review]

**Comment by muffato** [comment]
I've merged `master` into this branch, resolved the conflicts, and added nf-test tests.

**Review by maxulysse — APPROVED** [review]

**Pull request merged**